Source/RapyutaSimulationPlugins/Private/Drives/RobotVehicleMovementComponent.cpp

```diff
@@ -59,22 +74,52 @@ void URobotVehicleMovementComponent::UpdateOdom(float DeltaTime)
     {
         InitOdom();
     }
+
     // time
     float TimeNow = UGameplayStatics::GetTimeSeconds(GetWorld());
     OdomData.header_stamp_sec = static_cast<int32>(TimeNow);
     uint64 ns = (uint64)(TimeNow * 1e+09f);
     OdomData.header_stamp_nanosec = static_cast<uint32>(ns - (OdomData.header_stamp_sec * 1e+09));
 
+    // previous estimated data
+    FVector PreviousEstimatedPos = FVector(OdomData.pose_pose_position_x, OdomData.pose_pose_position_y, OdomData.pose_pose_position_z);
+    FQuat PreviousEstimatedRot = OdomData.pose_pose_orientation;
+
     // position
     FVector Pos = PawnOwner->GetActorLocation() - InitialTransform.GetTranslation();
+    FVector PreviousPos = PreviousTransform.GetTranslation();
+    PreviousTransform.SetTranslation(Pos);
+    Pos += PreviousEstimatedPos - PreviousPos + WithNoise * FVector(GaussianRNGPosition(Gen), GaussianRNGPosition(Gen), 0);
+
+    FRotator NoiseRot = FRotator(0, 0, WithNoise * GaussianRNGRotation(Gen));
+    FQuat Rot = InitialTransform.GetRotation().Inverse() * PawnOwner->GetActorRotation().Quaternion();
+    FQuat PreviousRot = PreviousTransform.GetRotation();
+    PreviousTransform.SetRotation(Rot);
+    Rot =  NoiseRot.Quaternion() * PreviousEstimatedRot * PreviousRot.Inverse() * Rot;
+    Rot.Normalize();
+
     OdomData.pose_pose_position_x = Pos.X;
     OdomData.pose_pose_position_y = Pos.Y;
     OdomData.pose_pose_position_z = Pos.Z;
-    OdomData.pose_pose_orientation = FQuat(PawnOwner->GetActorRotation() - InitialTransform.GetRotation().Rotator());
-
-    // velocity
-    OdomData.twist_twist_linear = Velocity;
-    OdomData.twist_twist_angular = FMath::DegreesToRadians(AngularVelocity);
+    OdomData.pose_pose_orientation = Rot;
+
+    // velocity - are signs correct?
+    OdomData.twist_twist_linear = (Pos - PreviousEstimatedPos)/DeltaTime;

```

Comment:
For linear velocity, in my code, I took into account the orientation modification, something like this:
`FVector NewVelocity = OdomData.pose_pose_orientation.UnrotateVector( Pos-PreviousEstimatedPos ) / DeltaTime;`
so that I can roughly get back the linear and angular velocities that were set with some ros twist msg.
But I don't know if this is the expected way to compute linear velocity for odom msg.

Reply:
changed according to your suggestion

it basically depends on the frame of reference that the twist_linear component should be expressed in.
Mine was in global coordinates, while what you suggest is in the frame of reference of the agent, so it should be mostly u=U, y=0, z=0 (where U is the cmd_vel velocity) for vehicles that use differential driving

@yuokamoto what frame of reference is needed?

Reply:
I'm adding the gazebo version for the differential drive, while leaving the other as default